src/client/websocket/handlers/TYPING_START.js

```diff
@@ -5,14 +5,63 @@ const { Events } = require('../../../util/Constants');
 module.exports = (client, { d: data }) => {
   const channel = client.channels.cache.get(data.channel_id);
   const user = client.users.cache.get(data.user_id);
+  const timestamp = new Date(data.timestamp * 1000);
 
   if (channel && user) {
-    /**
-     * Emitted whenever a user starts typing in a channel.
-     * @event Client#typingStart
-     * @param {Channel} channel The channel the user started typing in
-     * @param {User} user The user that started typing
-     */
-    client.emit(Events.TYPING_START, channel, user);
+    if (channel.type === 'voice') {
+      client.emit(Events.WARN, `Discord sent a typing packet to a voice channel ${channel.id}`);
+      return;
+    }
+
+    if (channel._typing.has(user.id)) {
+      const typing = channel._typing.get(user.id);
+
+      typing.lastTimestamp = timestamp;
+      typing.resetTimeout(tooLate(channel, user));
+    } else {
+      channel._typing.set(user.id, new TypingData(client, timestamp, timestamp, tooLate(channel, user)));
+
+      /**
+       * Emitted whenever a user starts typing in a channel.
+       * @event Client#typingStart
+       * @param {Channel} channel The channel the user started typing in
+       * @param {User} user The user that started typing
+       */
+      client.emit(Events.TYPING_START, channel, user);
+    }
   }
 };
+
+class TypingData {
+  constructor(client, since, lastTimestamp, _timeout) {
+    this.client = client;
+
+    this.since = since;
+
+    this.lastTimestamp = lastTimestamp;
```

Comment:
seeing as both of these will be the same on initalization, would it be better do
```js
this.since = this.lastTimestamp = since;
```
and just cutting out the `lastTimestamp` parameter from the constructor?

Reply:
Yes. It would be better to merge them into a single value at initialization. I've done that, and set `client` as a readonly property of the `TypingData` class.